modules/post/linux/manage/adduser.rb

```diff
@@ -0,0 +1,299 @@
+    groups_missing = get_missing_groups(group_file, groups)
+    unless groups_missing.empty?
+      if datastore['MissingGroups'] == 'ERROR'
+        fail_with(Failure::NotFound, "groups [#{groups_missing.join(' ')}] do not exist on the system. Change the `MissingGroups` Option to deal with errors automatically")
+      end
+      print_warning("Groups [#{groups_missing.join(' ')}] do not exist on system")
+      if datastore['MissingGroups'] == 'IGNORE'
+        groups -= groups_missing
+        print_good("Removed #{groups_missing.join(' ')} from target groups")
+      end
+    end
+
+    # Creating new groups if it was set and isnt manual
+    if groups.any? && datastore['MissingGroups'] == 'CREATE' && datastore['UseraddMethod'] != 'MANUAL'
+      # Since command can add on groups, checking over groups
+      fail_with(Failure::NotFound, 'Neither groupadd nor addgroup exist on the system. Try running with UseraddMethod as MANUAL to get around this issue') unless check_command_exists?('groupadd') || check_command_exists?('addgroup')
+      groupadd = check_command_exists?('groupadd') ? 'groupadd' : 'addgroup'
+
+      groups_missing.each do |group|
+        d_cmd_exec("#{groupadd} #{group}")
+        print_good("Added #{group} group")
+      end
+    end
+    groups
+  end
+
+  def run
+    fail_with(Failure::NoAccess, 'Session isnt running as root') unless is_root?
+    case datastore['UseraddMethod']
+    when 'CUSTOM'
+      fail_with(Failure::NotFound, "Cannot find command on path given: #{datastore['UseraddBinary']}") unless check_command_exists?(datastore['UseraddBinary'])
+    when 'AUTO'
+      fail_with(Failure::NotVulnerable, 'Cannot find a means to add a new user') unless check_command_exists?('useradd') || check_command_exists?('adduser')
+    end
+    fail_with(Failure::NotVulnerable, 'Cannot add user to sudo as sudoers doesnt exist') unless datastore['SudoMethod'] != 'SUDO_FILE' || file_exist?('/etc/sudoers')
+    fail_with(Failure::NotFound, 'Shell specified does not exist on system') unless check_command_exists?(datastore['SHELL'])
+    fail_with(Failure::BadConfig, "Username [#{datastore['USERNAME']}] is not a legal unix username.") unless datastore['USERNAME'] =~ /^[a-z][a-z0-9_-]{0,31}$/
+
+    # Encrypting password ahead of time
+    passwd = case datastore['PasswordHashType']
+             when 'DES'
+               UnixCrypt::DES.build(datastore['PASSWORD'])
+             when 'MD5'
+               UnixCrypt::MD5.build(datastore['PASSWORD'])
+             when 'SHA256'
+               UnixCrypt::SHA256.build(datastore['PASSWORD'])
+             when 'SHA512'
+               UnixCrypt::SHA512.build(datastore['PASSWORD'])
+             end
+
+    # Adding sudo to groups if method is set to use groups
+    groups = datastore['GROUPS']&.split || []
+    groups += ['sudo'] if datastore['SudoMethod'] == 'GROUP'
+    group_file = read_file('/etc/group').to_s
+    groups = validate_groups(group_file, groups)
```

Comment:
I'm not sure if handling groups prior to create the user is the best solution. When `MissingGroups` is set to `CREATE` and if the module fails to create the user for whatever reason,  the groups will be created anyway, which is probably not what the operator wants.

Reply:
It makes sense. One thing that can be done for now is to notify the operator that the operation failed but groups were created and should be cleaned up manually.

Reply:
So Currently it will print if any of the commands come up with an error, but error handling isnt dealt with because of how many possible errors there can be. Do you think a simple warning print would suffice?

Reply:
And we should have our warning